interp/src/interpreter.rs

```diff
@@ -1,101 +1,114 @@
-    eval_assigns(&(group.borrow()).assignments, &env)
+    eval_assigns(&(group.borrow()).assignments, env, component)
 }
 
 // XXX(karen): I think it will need another copy of environment for each
 // iteration of assignment statements
 /// Evaluates a group's assignment statements in an environment.
+/// How this is done:
+/// First, a new write-to environment is cloned from the original read-only environment.
+/// For each clock cycle (until the group's done signal is high):
+/// Then, each assignment statement is checked for its done signal is high.
+/// If that statement's done signal is high:
+/// If the assignment is combinational, it is immediately evaluated result and stored in the write-to environment.
+/// If it is not combinational, then it is added to an update queue, to be evaluated at the end of the current clock cycle.
+/// This continues until the group's done signal is high.
 fn eval_assigns(
     assigns: &[ir::Assignment],
-    env: &Environment,
+    mut env: Environment,
+    component: String,
 ) -> FutilResult<Environment> {
+    let cid = ir::Id::from(component.clone());
+
     // Find the done signal in the sequence of assignments
     let done_assign = get_done_signal(assigns);
 
-    // e2 = Clone the current environment
+    // Clone the current environment
     let mut write_env = env.clone();
 
     // XXX: Prevent infinite loops. should probably be deleted later
     // (unless we want to display the clock cycle)?
     let mut counter = 0;
 
     // Filter out the assignment statements that are not only from cells.
+    // Reorder assignment statements??
     // XXX: for now, also excludes cells not in the env map
     let ok_assigns = assigns
         .iter()
         .filter(|&a| {
             !a.dst.borrow().is_hole()
                 // dummy way of making sure the map has the a.src cell
-                && env.get_cell(&get_cell_from_port(&a.src)).is_some()
-                && env.get_cell(&get_cell_from_port(&a.dst)).is_some()
+                && env.get_cell(&cid, &get_cell_from_port(&a.src)).is_some()
+                && env.get_cell(&cid, &get_cell_from_port(&a.dst)).is_some()
         })
         .collect::<Vec<_>>();
 
-    // While done_assign.src is 0 (we use done_assign.src because done_assign.dst is not a cell's port; it should be a group's port)
-    while write_env.get_from_port(&done_assign.src.borrow()) == 0 && counter < 5
+    // XXX(yoona): At the moment interpreter rejects direct assignment of 1 to the groups
+    // needs to be fixed
+    if write_env.get_from_port(&cid, &done_assign.src.borrow()) == 1 {
+        panic!("TODO: done[group]=1 this group woud but be evaluated ");
+    }
+
+    // While done_assign.src is 0
+    // (we use done_assign.src because done_assign.dst is not a cell's port; it should be a group's port
+
+    while write_env.get_from_port(&cid, &done_assign.src.borrow()) == 0
+        && counter < 5
     {
+        env = write_env.clone();
```

Comment:
repeated cloning seems suspect, esp if updating in a loop

Reply:
env is passed in as mutable so we should probably just update it instead of cloning